src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs

```diff
@@ -0,0 +1,1383 @@
+        {
+            _ = tokenDescriptor ?? throw LogHelper.LogArgumentNullException(nameof(tokenDescriptor));
+
+            if (LogHelper.IsEnabled(EventLogLevel.Warning))
+            {
+                if ((tokenDescriptor.Subject == null || !tokenDescriptor.Subject.Claims.Any())
+                    && (tokenDescriptor.Claims == null || !tokenDescriptor.Claims.Any()))
+                    LogHelper.LogWarning(
+                        LogMessages.IDX14114, LogHelper.MarkAsNonPII(nameof(SecurityTokenDescriptor)), LogHelper.MarkAsNonPII(nameof(SecurityTokenDescriptor.Subject)), LogHelper.MarkAsNonPII(nameof(SecurityTokenDescriptor.Claims)));
+            }
+
+            if (tokenDescriptor.AdditionalHeaderClaims?.Count > 0 && tokenDescriptor.AdditionalHeaderClaims.Keys.Intersect(JwtTokenUtilities.DefaultHeaderParameters, StringComparer.OrdinalIgnoreCase).Any())
+                throw LogHelper.LogExceptionMessage(
+                    new SecurityTokenException(
+                        LogHelper.FormatInvariant(
+                            LogMessages.IDX14116,
+                            LogHelper.MarkAsNonPII(nameof(tokenDescriptor.AdditionalHeaderClaims)),
+                            LogHelper.MarkAsNonPII(string.Join(", ", JwtTokenUtilities.DefaultHeaderParameters)))));
+
+            if (tokenDescriptor.AdditionalInnerHeaderClaims?.Count > 0 && tokenDescriptor.AdditionalInnerHeaderClaims.Keys.Intersect(JwtTokenUtilities.DefaultHeaderParameters, StringComparer.OrdinalIgnoreCase).Any())
+                throw LogHelper.LogExceptionMessage(
+                    new SecurityTokenException(
+                        LogHelper.FormatInvariant(
+                            LogMessages.IDX14116,
+                            LogHelper.MarkAsNonPII(nameof(tokenDescriptor.AdditionalInnerHeaderClaims)),
+                            LogHelper.MarkAsNonPII(string.Join(", ", JwtTokenUtilities.DefaultHeaderParameters)))));
+
+            return CreateToken(
+                tokenDescriptor,
+                SetDefaultTimesOnTokenCreation,
+                TokenLifetimeInMinutes);
+        }
+
+        internal static string CreateToken(
+            SecurityTokenDescriptor tokenDescriptor,
+            bool setdefaultTimesOnTokenCreation,
+            int tokenLifetimeInMinutes)
+        {
+            // The form of a JWS is: Base64UrlEncoding(UTF8(Header)) | . | Base64UrlEncoding(Payload) | . | Base64UrlEncoding(Signature)
+            // Where the Header is specifically the UTF8 bytes of the JSON, whereas the Payload encoding is not specified, but UTF8 is used by everyone.
+            // The signature is over ASCII(Utf8Bytes(Base64UrlEncoding(Header) | . | Base64UrlEncoding(Payload)))
+            // Since it is not known how large the JWS will be, a MemoryStream is used.
+            // An ArrayBufferWriter was benchmarked, while slightly faster, more memory is used and different code would be needed for 461+ and net6.0+
```

Comment:
Probably for future work, but a IBufferWriter would fit better if we ever want to do a PooledByteBuffer

Reply:
@keegan-caruso yes, .net 8 will make that public.